plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java

```diff
@@ -0,0 +1,233 @@
+                    tableName,
+                    tableLocation,
+                    Optional.ofNullable(metadataFileName));
+        }
+    }
+
+    private void doRegisterTable(
+            ConnectorSession clientSession,
+            String schemaName,
+            String tableName,
+            String tableLocation,
+            Optional<String> metadataFileName)
+    {
+        if (!registerTableProcedureEnabled) {
+            throw new TrinoException(PERMISSION_DENIED, "register_table procedure is disabled");
+        }
+        checkProcedureArgument(schemaName != null && !schemaName.isEmpty(), "schema_name cannot be null or empty");
+        checkProcedureArgument(tableName != null && !tableName.isEmpty(), "table_name cannot be null or empty");
+        checkProcedureArgument(tableLocation != null && !tableLocation.isEmpty(), "table_location cannot be null or empty");
+        metadataFileName.ifPresent(RegisterTableProcedure::validateMetadataFileName);
+
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+        TrinoCatalog catalog = catalogFactory.create(clientSession.getIdentity());
+        if (!catalog.namespaceExists(clientSession, schemaTableName.getSchemaName())) {
+            throw new TrinoException(SCHEMA_NOT_FOUND, format("Schema '%s' does not exist", schemaTableName.getSchemaName()));
+        }
+
+        TrinoFileSystem fileSystem = fileSystemFactory.create(clientSession);
+        String metadataLocation = getMetadataLocation(fileSystem, tableLocation, metadataFileName);
+        validateLocation(fileSystem, metadataLocation);
+        try {
+            // Try to read the metadata file. Invalid metadata file will throw the exception.
+            TableMetadataParser.read(fileSystem.toFileIo(), metadataLocation);
+        }
+        catch (RuntimeException e) {
+            throw new TrinoException(ICEBERG_INVALID_METADATA, metadataLocation +" is not a valid metadata file", e);
+        }
+
+        catalog.registerTable(clientSession, schemaTableName, tableLocation, metadataLocation);
+    }
+
+    private static void validateMetadataFileName(String fileName)
+    {
+        String metadataFileName = fileName.trim();
+        checkProcedureArgument(!metadataFileName.isEmpty(), "metadata_file_name cannot be empty when provided as an argument");
+        checkProcedureArgument(!metadataFileName.contains("/"), "%s is not a valid metadata file", metadataFileName);
+    }
+
+    /**
+     * Get the latest metadata file location present in location if metadataFileName is not provided, otherwise
+     * form the metadata file location using location and metadataFileName
+     */
+    private static String getMetadataLocation(TrinoFileSystem fileSystem, String location, Optional<String> metadataFileName)
+    {
+        return metadataFileName
+                .map(name -> format("%s/%s/%s", stripTrailingSlash(location), METADATA_FOLDER_NAME, name.trim()))
```

Comment:
why `name.trim()`?

Reply:
To handle the case when user has provided the filename with prefix whitespace or trailing whitespace. 

But I think metadata file validation will take of this case by throwing exceptions. So removed the `name.trim()` now.